packages/middleware/src/loggerManager.ts

```diff
@@ -0,0 +1,108 @@
+import type { Response } from "express";
+import { LoggerInterface } from "@vue-storefront/logger";
+import type {
+  AlokaiContainer,
+  IntegrationContext,
+  MiddlewareConfig,
+} from "./types";
+
+type LoggerInstance = LoggerInterface;
+const GLOBAL_MIDDLEWARE_CFG_KEY = Symbol("GLOBAL_MIDDLEWARE_CFG_KEY");
+export class LoggerManager<
+  TLoggerConfig = any,
+  TIntegrationContext extends Record<string, IntegrationContext> = any
+> {
+  private instances: Record<string | symbol, LoggerInstance> = {};
+
+  /** Creating minimal amount of needed Logger instances * */
+  constructor(
+    config: MiddlewareConfig<TIntegrationContext>, // Whole middleware.config.js object
+    buildLogger: (logger: TLoggerConfig) => LoggerInstance
+  ) {
+    // Global Logger
+    this.instances[GLOBAL_MIDDLEWARE_CFG_KEY] = buildLogger(
+      config.logger as TLoggerConfig // logger is on same level as integrations
+    );
+    // Logger for every integration having custom logger config
+    for (const [integrationName, integrationEntry] of Object.entries(
+      config.integrations
+    )) {
+      if (this.hasCustomLoggerConfig(integrationEntry)) {
+        // every integration inside `configuration` object can have own logger too
+        this.instances[integrationName] = buildLogger(
+          integrationEntry.logger as TLoggerConfig
+        );
+      }
+    }
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  private hasCustomLoggerConfig(integrationEntry) {
+    return Boolean(integrationEntry.logger);
+  }
+
+  // getting reference to instance of logger by integration tag name
+  public get(integrationName) {
+    return (
+      this.instances[integrationName] ||
+      this.instances[GLOBAL_MIDDLEWARE_CFG_KEY]
+    );
+  }
+
+  // getting reference to global instance of logger
+  public getGlobal() {
+    return this.instances[GLOBAL_MIDDLEWARE_CFG_KEY];
+  }
+}
+
+export function getLogger(
+  source: AlokaiContainer | Response | { res: Response }
+): LoggerInstance {
+  if (!source) {
+    throw new Error("Parameter for getLogger function not provided");
+  }
+  if ("logger" in source) {
+    return source.logger;
+  }
+  let base: Response;
+  if ("res" in source) {
+    base = source.res;
+  } else {
+    base = source;
+  }
```

Comment:
```suggestion
 const base = source?.res ?? source;
```

Reply:
TypeScript won't allow that (accessing `.res`). `"res" in source` is a way of selecting type from types intersection that TypeScript understands. If you really want to make it one liner, it could be:
```ts
const base = "res" in source ? source?.res : source;
return base.locals.alokai.logger;
```

But I also find it a little inelegant. Maybe smth like that?:
```ts
type ContextSubset = { res: Response };
type LoggerSource = AlokaiContainer | Response | ContextSubset;
function isAlokaiContainer(source: LoggerSource): source is AlokaiContainer {
  return "logger" in source;
}
function isContext(source: LoggerSource): source is ContextSubset {
  return "res" in source;
}

export function getLogger(source: LoggerSource): LoggerInterface {
  if (isAlokaiContainer(source)) {
    return source.logger;
  }
  const base = isContext(source) ? source.res : source;
  return base.locals.alokai.logger;
}
```

Reply:
Also, as a part of other tip, getLogger would be moved to another dedicated file, and only getLogger would be exported. So it's functions won't decrease readability of the logger manager's file.

Reply:
both code versions are better than initial one so feel free to pick whatever you like more :)